all: use go1.18

keegancsmith opened the pull request:
This bumps go.mod, sg and asdf to use the latest version of go. Other commits have updated us to go1.18, but this is the important one since it affects both our dev env and CI.

Brew ships go 1.18.1 so this should be good to go.

### Test Plan

Ran "sg start oss". main dry run on CI.

eseliger commented:
Oh no that's nasty :( It fails, because the `compare-hash.sh` script checks if the executor binary changes compared to the previous commit. On that commit, there's go 1.17 as a requirement, but we don't run `asdf install` again .. I think we could add `asdf install` to the script, but not sure if that has any side effects? 

cc @bobheadxi who wrote the initial compare hash script I believe.

keegancsmith commented:
> Oh no that's nasty :( It fails, because the `compare-hash.sh` script checks if the executor binary changes compared to the previous commit. On that commit, there's go 1.17 as a requirement, but we don't run `asdf install` again .. I think we could add `asdf install` to the script, but not sure if that has any side effects?

But go1.18 will produce a different binary to go1.17? What is this supposed to protect against?

> > I think we could add asdf install to the script, but not sure if that has any side effects?
> 
> I think we can do that! IIRC a similar thing was added by @efritz for the backcompat tests, which also runs into tool incompatibilities
> 
> https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/dev/ci/go-backcompat/test.sh?L133-136&utm_campaign=search&utm_source=raycast-sourcegraph

I'll push a commit to see if this helps, but I'm a bit confused to what we are testing here TBH.

bobheadxi commented:
`compare-hash.sh` _soft-fails_ on a diff and uses that to indicate that executor should be rebuilt and published (conditional because executor takes forever to build)

That said, @keegancsmith I think you are right - I don't think you need to take any action here on that front, it's expected that there is a diff

keegancsmith commented:
> `compare-hash.sh` _soft-fails_ on a diff and uses that to indicate that executor should be rebuilt and published (conditional because executor takes forever to build)
> 
> That said, @keegancsmith I think you are right - I don't think you need to take any action here on that front, it's expected that there is a diff. Looking at the build logs https://buildkite.com/sourcegraph/sourcegraph/builds/145774#2e65455b-fa4c-4aec-87d2-c8a5686e5960 everything looks fine?

Ok, to get confidence I am going to do `[skip-hash-compare]` in msg, that seems to do what I want. Then once it is green I'll merge in without that skip?

eseliger commented:
I think the build on main will fail then, but only once, the next commits should pass again. 

Maybe we should just add the line @bobheadxi recommended to the script?

keegancsmith commented:
> Maybe we should just add the line @bobheadxi recommended to the 
> script?

I did and it still failed. See above. Me adding the skip thing has made
the CI proceed further.

eseliger commented:
Seems like the executors build passed now?

keegancsmith commented:
CI is green! I've now pulled out the symbols change into https://github.com/sourcegraph/sourcegraph/pull/34998. I rebased this branch onto that and pushed to see if it stays green. I'll wait until that PR lands, then will rebase again and open this up for review. @bobheadxi for when this is reviewed and merged into main, do I keep `[skip-hash-compare]` on the PR, but then remove it from the main commit message?

@eseliger I just took this branch, removed `[skip-hash-compare]`, cherry-picked that commit and pushed it to https://github.com/sourcegraph/sourcegraph/compare/main-dry-run/k/eseliger-executor-fun. You should be able to check the CI status there.

eseliger commented:
@keegancsmith that one went green :)

keegancsmith commented:
woohoo. Pulled it out into https://github.com/sourcegraph/sourcegraph/pull/35002. Once that lands I'll rebase and open this up for review. But everything is looking green. Are you ready for people to destroy our codebase with generics?

efritz commented:
> Are you ready for people to destroy our codebase with generics?

NO.

bobheadxi reviewed (approved):
> Are you ready for people to destroy our codebase with generics?

im ready!

eseliger reviewed (approved):
niceee

camdencheek reviewed (approved):
🥳
